### PR TITLE
[Draft] Adding commit hash support.

### PR DESCRIPTION
**Pull request by hiragram**
Hi. First of all, thank you so much for your great work on Mint!

## Description

I want to specify one exact commit rather than branch or tag name in Mintfile. 

And I've found this line: 

https://github.com/yonaskolb/Mint/blob/6d889541bf3c8935bad858e15dee3937d5cc119c/Sources/MintKit/Mintfile.swift#L41

When I saw this line, I thought Mintfile already support specifying version using commit hash like `yonaskolb/Mint@9010ebe` but it doesn't work because because `git clone -b` doesn't support cloning using commit hash.

ref: https://github.com/yonaskolb/Mint/blob/784ecd0d158b25c88e87101977af163fd21cf8ed/Sources/MintKit/Mint.swift#L248

I've experimentally added commit hash support like `yonaskolb/Mint@commit:9010ebe` on this PR.
It's still draft because no tests are added and I haven't understood Mint completely.
These changes are very rough but you can review my concept.
Do you have any ideas about this?

## Questions

- When does this code executed?  In my PR, commit hash is specified as `yonaskolb/Mint@commit:9010ebe` so it conflicts to these lines, but I don't understand when `:` appeares after `@` in current specification or implementation.
https://github.com/yonaskolb/Mint/blob/d955d04edddd9ecb0ce6701395c8e740e30bf9d6/Sources/MintKit/PackageReference.swift#L23-L25 

- [just curious] Are there any usecases that `packageParts.count == 3` becomes `true`? I don't know when 2 `@`s appear in one line.
https://github.com/yonaskolb/Mint/blob/d955d04edddd9ecb0ce6701395c8e740e30bf9d6/Sources/MintKit/PackageReference.swift#L19-L21

### CODE REVIEW

**Comment by hiragram** [comment]
@dchohfi hello.
This is still draft and very conceptual. I want to hear what @yonaskolb thinks about this design before I start implementing properly. So I have no more progress than this PR.

File: Sources/MintKit/PackageReference.swift [hunk 1]
```diff
@@ -73,6 +118,6 @@ public class PackageReference {
 
 extension PackageReference: Equatable {
     public static func == (lhs: PackageReference, rhs: PackageReference) -> Bool {
-        return lhs.repo == rhs.repo && lhs.version == rhs.version
+        return lhs.repo == rhs.repo && lhs.version.string == rhs.version.string
```
Comment (yonaskolb) [inline Sources/MintKit/PackageReference.swift]: version is already equatable

File: Sources/MintKit/PackageReference.swift [hunk 2]
```diff
@@ -3,9 +3,54 @@ import PathKit
...
+            let parts = versionString.split(separator: ":")
+
+            if parts.count == 1 {
+                self = .branch(name: versionString)
+            } else if parts.count == 2 {
+                let name = String(parts[1])
+                switch parts[0] {
+                case "tag":
+                    self = .tag(name: name)
+                case "commit":
```
Comment (yonaskolb) [inline Sources/MintKit/PackageReference.swift]: can we add "branch" as well

File: Sources/MintKit/PackageReference.swift [hunk 3]
```diff
@@ -3,9 +3,54 @@ import PathKit
...
+            case .unspecified:
+                return ""
+            }
+        }
+
+        init(versionString: String) {
+            let parts = versionString.split(separator: ":")
+
+            if parts.count == 1 {
+                self = .branch(name: versionString)
```
Comment (yonaskolb) [inline Sources/MintKit/PackageReference.swift]: Can we add a new case `unknown`. If the string has 2 dots in it though let's assume it's `.tag`

File: Sources/MintKit/PackageReference.swift [hunk 4]
```diff
@@ -3,9 +3,54 @@ import PathKit
...
+            } else if parts.count == 2 {
+                let name = String(parts[1])
+                switch parts[0] {
+                case "tag":
+                    self = .tag(name: name)
+                case "commit":
+                    self = .commit(hash: name)
+                default:
+                    self = .unspecified
```
Comment (yonaskolb) [inline Sources/MintKit/PackageReference.swift]: This can be `.unknown`

File: Sources/MintKit/PackageReference.swift [hunk 5]
```diff
@@ -3,9 +3,54 @@ import PathKit
...
 public class PackageReference {
     public var repo: String
-    public var version: String
+    public var version: Revision
 
-    public init(repo: String, version: String = "") {
+    public enum Revision: Equatable {
+        case tag(name: String)
+        case branch(name: String)
+        case commit(hash: String)
```
Comment (yonaskolb) [inline Sources/MintKit/PackageReference.swift]: Can we remove the argument labels

File: Sources/MintKit/PackageReference.swift [hunk 6]
```diff
@@ -3,9 +3,54 @@ import PathKit
 
 public class PackageReference {
     public var repo: String
-    public var version: String
+    public var version: Revision
 
-    public init(repo: String, version: String = "") {
+    public enum Revision: Equatable {
```
Comment (yonaskolb) [inline Sources/MintKit/PackageReference.swift]: Lets make this conform to `ExpressibleByStringLiteral`

File: Tests/MintTests/MintTests.swift [hunk 7]
```diff
@@ -121,7 +121,7 @@ class MintTests: XCTestCase {
 
         try mint.bootstrap()
 
-        let package = PackageReference(repo: "yonaskolb/SimplePackage", version: "4.0.0")
+        let package = PackageReference(repo: "yonaskolb/SimplePackage", version: PackageReference.Revision.branch(name: "4.0.0"))
```
Comment (yonaskolb) [inline Tests/MintTests/MintTests.swift]: Once Revision has `ExpressibleByStringLiteral` conformance, these can all stay the same

File: Sources/MintKit/Mint.swift [hunk 8]
```diff
@@ -248,7 +248,17 @@ public class Mint {
...
         try? packageCheckoutPath.delete()
 
-        let cloneCommand = "git clone --depth 1 -b \(package.version) \(package.gitPath) \(package.repoPath)"
+        let cloneCommand: String
+
+        switch package.version {
+        case .branch(name: let name), .tag(name: let name):
+            cloneCommand = "git clone --depth 1 -b \(name) \(package.gitPath) \(package.repoPath)"
+        case .commit(hash: let hash):
+            cloneCommand = "git clone \(package.gitPath) \(package.repoPath) && cd \(package.repoPath) && git checkout \(hash)"
```
Comment (yonaskolb) [inline Sources/MintKit/Mint.swift]: We can use this clone and checkout for `.unknown` as well to be safe

**Comment by hiragram** [comment]
Hi @yonaskolb . Thank you for your review.
In your plan, I am worried that branches named `develop` will be involved. Does it sound good to you? I have no idea how many people will get affected and how much execution will get slower.

I also understood your answers to my questions. In my opinion, it's better to stop using symbols like `@` to specify repository and/or revision. If Mintfile is designed like `github yonaskolb/Mint commit:9010ebe` rather than `yonaskolb/Mint@9010ebe`, things can get more easier, I think. This is extremely breaking change, but there are no chance to do that after Mint has reached to version 1 . I understand it's not the main subject on this PR. So let me open another issue if you think there is room for discussion.

**Comment by yonaskolb** [comment]
To mitigate any slowness when using branches, some extra heuristics could be employed:
- checking for explicit common names like `master`, `develop` or `dev`
- assuming that if a string is still 7 or more chars, but doesn't have numbers in it, it must no be a char. That is not sound proof and could lead to false positives, though I believe it's HIGHLY unlikely that a SHA doesn't have a number in it. If that were ever to happen a user could enter a longer sha which would have a number in it
In the end, having to clone the whole repo isn't the worst thing anyway.

I agree about `@` not being a perfect way to specify a version. It's just been there since the beginning and parsing `git@x` and `ssh://..` came after and complicated things. I recommend opening a new issue for that for discussion

**Comment by hiragram** [comment]
Got it.
I am opening another issue to discuss about re-design of Mintfile.
This PR strongly depends on design of Mintfile, so I keep this PR pending until the discussion has completed.

**Comment by yonaskolb** [comment]
Closing this in favour of #172

**Pull request closed without merging**